posts/fantastic-simprocs.md

```diff
@@ -0,0 +1,198 @@
+link: ''
+slug: fantastic-simprocs
+tags: 'simp, simproc, meta'
+title: Fantastic Simprocs and Where to Find Them.
+type: text
+---
+
+Lean core added support for custom simplification procedures, aka *simprocs*.
+This blog post is the first in a series of two aimed at explaining what a simproc is, what kind of problems can be solved with simprocs, and what tools we have to write them.
+
+<!-- TEASER_END -->
+
+This post describes purely informally what simprocs are and do.
+The second post will be a walkthrough to writing a simple simproc in three different ways.
+
+To understand what a simproc is and how it works, we will first explain how `simp` works.
+Then we will give some examples and non-examples of simprocs.
+
+# How simp works
+
+`simp` is made of two components.
+
+The first component is **rewriting rules**.
+Almost all rewriting rules are lemmas that prove an equality and are tagged with `@[simp]` in eg Lean or Mathlib.
+A rewriting rule is characterised by its *left hand side* and *right hand side*.
+Eg for a lemma of the form `LHS = RHS`, this is `LHS` and `RHS` respectively.
+
+The second one is the **`simp` tactic**, which we will refer to simply as `simp`.
+When run on a goal `⊢ e`, `simp` iteratively looks for a subexpression of `e` that matches the left hand side of some rewriting rule, and replaces that subexpression with the right hand side of that rule.
+
+For example, here's the proof steps `simp` follows to close the goal `37 * (Nat.fib 0 + 0) = 0`
+```lean
+⊢ 37 * (Nat.fib 0 + 0) = 0
+⊢ 37 * (0 + 0) = 0
+⊢ 37 * 0 = 0
+⊢ 0 = 0
+⊢ True
+```
+
+> The order in which `simp` traverses an expression is relatively intricate.
+> As a simple approximation, simplification is performed from the inside-out.
+> See the next blog post for full details on the algorithm.
+
+> If you write `set_option trace.Meta.Tactic.simp true in example : MyGoal := by simp`, you will see the list of simplification steps `simp` performs on `Goal`.
+
+# What is a simproc?
+
+Earlier, we said "almost all rewriting rules are lemmas".
+What about rewriting rules that are not lemmas?
+This is where simprocs enter the picture.
+
+A simproc is a program which, given an expression `LHS`, computes a new expression `RHS` and finds a proof of `LHS = RHS` on the fly.
+
+The concept of a simproc is genuinely more powerful than that of a simp lemma.
+Even though the right hand side of a lemma changes depending on its left hand side, it only does so via *syntactic substitution*.
+
+For example, the lemma `Nat.mul_add (a b c : Nat) : a * (b + c) = a * b + a * c` can be specialized to have right hand side `0 * b + 0 * c`, `a * b + a * (c + d)`, etc...
+But all these have *shape* `_ * _ + _ * _` and eg no value of `a`, `b`, `c` will make it have shape `_ * _ + _ * _ + _ * _`.
+
+With this in mind, simprocs are *parametric simp lemmas*.
+
+## Everything is a simproc
+
+As a very neat unification of concepts, it turns out that `simp` only ever handles simprocs.
```

Comment:
```suggestion
As a very neat unification of concepts, it turns out that the implementation of `simp` only ever handles simprocs.
```

Reply:
Hmm, your suggestion sounds a bit clunky

Reply:
The distinction I want to draw here is that you first said `simp` is made of two components, then you're saying that actually the first one was a lie. And so adding "implementation" here makes it clearer that these are consistent viewpoints, one is how simp intuitively works and the other is how it's practically implemented. I'm happy with alternate phrasing to a similar effect, but I think the current phrasing gives a confusing message here.